onionshare/web/receive_mode.py

```diff
@@ -214,19 +153,30 @@ def __call__(self, environ, start_response):
         return self.app(environ, start_response)
 
 
-class ReceiveModeTemporaryFile(object):
+class ReceiveModeFile(object):
     """
-    A custom TemporaryFile that tells ReceiveModeRequest every time data gets
-    written to it, in order to track the progress of uploads.
+    A custom file object that tells ReceiveModeRequest every time data gets
+    written to it, in order to track the progress of uploads. It starts out with
+    a .part file extension, and when it's complete it removes that extension.
     """
     def __init__(self, request, filename, write_func, close_func):
         self.onionshare_request = request
         self.onionshare_filename = filename
         self.onionshare_write_func = write_func
         self.onionshare_close_func = close_func
 
-        # Create a temporary file
-        self.f = tempfile.TemporaryFile('wb+')
+        self.filename = os.path.join(self.onionshare_request.receive_mode_dir, filename)
+        self.filename_in_progress = '{}.part'.format(self.filename)
+
+        # Open the file
+        self.upload_error = False
+        try:
+            self.f = open(self.filename_in_progress, 'wb+')
+        except:
+            # This will only happen if someone is messaging with the data dir while
+            # OnionShare is running, but if it does make sure to throw an error
+            self.upload_error = True
+            self.f = tempfile.TemporaryFile('wb+')
```

Comment:
Do you mean to deliberately still use a TemporaryFile as a sort of fallback here, even though you're considering this an upload error? Should it not simply return early?

Reply:
I actually do do that deliberately. The reason is because `self.f` needs to be a file-like object. Originally I was setting to to `None`, but that causes another exception to get thrown later on, when it tries to call `.write` on it (which it will try to call at least once, before it can cancel gracefully).